### ISSUE TITLE
Compiler crashes with `Inconsistent state in TS` assertion

### ISSUE THREAD

**Issue by vpatryshev** [lampepfl/dotty#12839]
## Compiler version

3.0.1-RC1

## Minimized code

```Scala
class Sample {
  def sample: Unit = {
    def set: Set[X] = ???
    val list: List[X] = Nil
    for { _ <- set } yield list.toMap
  }
}
```

## Output (click arrow to expand)
<details>

```scala
Vlad4087688721:Categories vpatryshev$ sbt "clean;compile" > err.log 2>&1
Vlad4087688721:Categories vpatryshev$ ls -l err.log
-rw-r--r--  1 vpatryshev  staff  24908 Jun 15 21:11 err.log
Vlad4087688721:Categories vpatryshev$ less err.log





















[error]         at sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2319)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:46)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:28)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:23)
[error]         at sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error]         at sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2315)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error]         at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error]         at sbt.Execute.work(Execute.scala:291)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error]         at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error]         at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error]         at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error]         at java.base/java.lang.Thread.run(Thread.java:834)
[error] (scala3 / Compile / compileIncremental) java.lang.AssertionError: assertion failed: Inconsistent state in TS[10X, 2, 1, 0]: it owns TypeVar(TypeParamRef(K)) whose owningState is TS[2, 1, 0]
[error] Total time: 4 s, completed Jun 15, 2021, 9:11:59 PM
```
</details>

**Comment by som-snytt** [comment]
What is `X`?

**Comment by griggt** [comment]
Duplicate of #12736. Fixed by #12827.

**Comment by vpatryshev** [comment]
@som-snytt absolutely anything. I just minimized the code.
